### PR TITLE
Deprecate `skimage.morphology.skeletonize_3d` in favor of just `skeletonize`

### PR DESCRIPTION
**Pull request by lagru**
## Description

Personally I wouldn't mind deprecating the `skeletonize` wrapper altogether and thereby making the API more explicit and maintenance easier. But that might be more controversial.

See also #7090.

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
...
```

### CODE REVIEW

**Comment by jni** [comment]
@lagru personally I think we should only have one skeletonize function, and we should instead deprecate skeletonize_3d. This is because we can unambiguously decide between 2D and 3D based on the input array. It was meant to be a followup to #3904, but I never followed through on it. 😬

**Comment by mkcor** [comment]
+1 to deprecating `skeletonize_3d`

File: skimage/morphology/_skeletonize.py [hunk 1]
```diff
@@ -159,13 +160,25 @@ def skeletonize_2d(image):
...
+def skeletonize_2d(image):
+    return _skeletonize_2d(image)
+
+
+skeletonize_2d.__doc__ = _skeletonize_2d.__doc__
+skeletonize_2d = deprecate_func(
+    deprecated_version="0.22",
+    removed_version="0.24",
+    hint="Use `skimage.morphology.skeletonize` instead."
+)(skeletonize_2d)
```
Comment (jni) [inline skimage/morphology/_skeletonize.py:179]: Why this instead of using `deprecate_func` as a decorator...?

**Review by jni — APPROVED** [review]
@lagru sorry that I missed your updates here. Can you fix the merge conflicts? I also had a question about the verbosity of using the deprecate_func decorator, but that's not blocking.

**Comment by jni** [comment]
Interestingly, the bot doesn't count reviews as comments! 🤔

**Comment by jni** [comment]
... and/or it's not autoremoving the label on PRs for some reason. I'm gonna remove it manually LOL.

File: skimage/morphology/_skeletonize.py (hunk 1 above)
Comment (lagru) [reply skimage/morphology/_skeletonize.py:179]: `deprecate_func` modifies the docstring and I didn't copy over the docstring to the placeholder `skeletonize_2d` so I'm doing that at runtime before wrapping. I'm not sure why I did it that way but maybe to keep the diff smaller and preserver history for the original docstring.. I can change it if you prefer the other approach.

**Comment by lagru** [comment]
> Interestingly, the bot doesn't count reviews as comments! 🤔

It only [runs once each day](https://github.com/scikit-image/scikit-image/blob/c3bd6725ae42578287037ca13c60fb57d75ee651/.github/workflows/dormant_issues_prs.yml#L7), so that's probably why it didn't react to your activity yet.

**Comment by lagru** [comment]
I'm somewhat unhappy with the current `skeletonize` codebase. The "historic" test structure with 3D stuff being in its own file makes what's being tested / covered more opaque. And handling user input in a tolerant way that also minimizes unnecessary copies feels to subtle for my taste. But it works and we have better things to do than refactor the whole thing.

@jni, I requested your review again as resolving the merge conflicts was more complex than usual and I tweaked a few minor things that seemed weird after the merge.

**Comment by lagru** [comment]
I haven't investigated thoroughly yet but I think the optional dependencies are causing at least some of the failures and are unrelated (also happening on `main`)..

File: skimage/morphology/_skeletonize.py [hunk 2]
```diff
@@ -100,9 +101,8 @@ def skeletonize_2d(image):
     Parameters
     ----------
     image : numpy.ndarray
-        A binary image containing the objects to be skeletonized. '1'
-        represents foreground, and '0' represents background. It
-        also accepts arrays of boolean values where True is foreground.
+        An image containing the objects to be skeletonized. Zeros
+        represent background, nonzero values are foreground.
```
Comment (mkcor) [inline skimage/morphology/_skeletonize.py:105]:
```suggestion
        An image containing the objects to be skeletonized. Zeros or ``False``
        represent background, nonzero values or ``True`` are foreground.
```
As lines 23-24?

File: skimage/morphology/_skeletonize.py [hunk 3]
```diff
@@ -595,7 +593,7 @@ def skeletonize_3d(image):
     Parameters
     ----------
     image : ndarray, 2D or 3D
-        A binary image containing the objects to be skeletonized. Zeros
+        An image containing the objects to be skeletonized. Zeros
```
Comment (mkcor) [inline skimage/morphology/_skeletonize.py:596]: Same question as above.

File: skimage/morphology/_skeletonize.py [hunk 4]
```diff
@@ -72,26 +73,26 @@ def skeletonize(image, *, method=None):
            [0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=uint8)
 
     """
-    image = image.astype(bool, copy=False)
+    image = image.astype(bool, order="C", copy=False)
```
Comment (mkcor) [inline skimage/morphology/_skeletonize.py:76]: Out of curiosity, what prompted this addition? Thanks!

**Review by mkcor — COMMENTED** [review]
Nice to see some progress wrt the [skimage2 transition list](https://github.com/scikit-image/scikit-image/wiki/API-changes-for-skimage2)!

**Comment by jni** [comment]
> It only [runs once each day](https://github.com/scikit-image/scikit-image/blob/c3bd6725ae42578287037ca13c60fb57d75ee651/.github/workflows/dormant_issues_prs.yml#L7), so that's probably why it didn't react to your activity yet.

Ah! I thought it would run whenever there was a new comment. 😅 All good.

File: skimage/morphology/_skeletonize.py (hunk 4 above)
Comment (jni) [reply skimage/morphology/_skeletonize.py:76]: I am also curious. 😂

Comment (lagru) [reply skimage/morphology/_skeletonize.py:76]: All code paths lead eiter to `_compute_thin_image` or `_fast_skeletonize` which both require C-contiguous memoryviews. This makes sure that the function doesn't fail unncessarily when users pass in a different layout. 

While it is currently a bit redundant with the line 
```python
image_o = image.astype(bool, order="C", copy=False)
```
in `_skeletonize_3d`, I thought this doesn't hurt robustness as in "defensive coding".

Comment (mkcor) [reply skimage/morphology/_skeletonize.py:76]: Sure, defensive is better than lax! Thanks for the details. But, in practice, have you run into a failure, or did you just notice the code paths and the current 'risk?'

**Pull request merged**